examples/Example6/example6.cu

```diff
@@ -304,6 +306,89 @@ __global__ void PerformStep(adept::BlockData<track> *allTracks, adept::MParray *
   }
 }
 
+// A parallel version of LoopNavigator::RelocateToNextVolume.
+__global__ void RelocateToNextVolume(adept::BlockData<track> *allTracks, adept::MParray *relocateQueue)
+{
+  int queueSize = relocateQueue->size();
+  // Note the loop setup: All threads in a block relocate one particle.
+  // for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < queueSize; i += blockDim.x * gridDim.x) {
+  for (int i = blockIdx.x; i < queueSize; i += gridDim.x) {
+    int particleIndex   = (*relocateQueue)[i];
+    track &currentTrack = (*allTracks)[particleIndex];
+
+    vecgeom::NavStateIndex &state = currentTrack.current_state;
+
+    __shared__ vecgeom::Precision localCoordinates[3];
+    __shared__ vecgeom::VPlacedVolume const *currentVolume;
+    if (threadIdx.x == 0) {
+      // Push the point inside the next volume.
+      vecgeom::Vector3D<vecgeom::Precision> pushed = currentTrack.pos + 1.E-6 * currentTrack.dir;
+
+      // Calculate local point from global point.
+      vecgeom::Transformation3D m;
+      state.TopMatrix(m);
+      vecgeom::Vector3D<vecgeom::Precision> localPoint = m.Transform(pushed);
+
+      currentVolume = state.Top();
+
+      while (currentVolume && (currentVolume->IsAssembly() || !currentVolume->UnplacedContains(localPoint))) {
+        state.Pop();
+        localPoint    = currentVolume->GetTransformation()->InverseTransform(localPoint);
+        currentVolume = state.Top();
+      }
+
+      localCoordinates[0] = localPoint.x();
+      localCoordinates[1] = localPoint.y();
+      localCoordinates[2] = localPoint.z();
+    }
+
+    __syncthreads();
+
+    if (currentVolume) {
+      bool godeeper;
+      do {
+        godeeper = false;
+
+        vecgeom::Vector3D<vecgeom::Precision> localPoint(localCoordinates[0], localCoordinates[1], localCoordinates[2]);
+
+        const auto &daughters = currentVolume->GetDaughters();
+        auto daughtersSize    = daughters.size();
+        vecgeom::Vector3D<vecgeom::Precision> transformedPoint;
+        vecgeom::VPlacedVolume const *nextVolume = nullptr;
+        for (int d = threadIdx.x; d < daughtersSize; d += blockDim.x) {
+          const auto *daughter = daughters[d];
+          if (daughter->Contains(localPoint, transformedPoint)) {
+            nextVolume = daughter;
+            break;
+          }
+        }
+
+        auto *previousVolume = currentVolume;
+        __syncthreads();
+        if (nextVolume) {
+          localCoordinates[0] = transformedPoint.x();
+          localCoordinates[1] = transformedPoint.y();
+          localCoordinates[2] = transformedPoint.z();
+
+          currentVolume = nextVolume;
+          state.Push(currentVolume);
```

Comment:
Note that this leads to undefined behavior in case two of the daughters are overlapping, and the propagated point is in the overlapping region. In the sequential version, always the same daughter wins. To have the same here maybe the `currentVolume` and `localCoordinates` need to be atomic.

Reply:
... but even this will not eliminate undefined behavior

Reply:
My assumption was that
1. a volume strictly contains all daughters, and
2. all daughters of a given volume are non-intersecting.

So you're saying that the second point is not correct? If so, what is the motivation for using the first daughter that contains a point? After all, using an acceleration structure might change ordering. Or do we just need to pick one (and always the same)?

Reply:
In fact, this happens only in the case of geometry definition errors, but each large LHC geometry has many of those. Picking any of the candidates in such a case is equivalent (both bad), what matters is that the order is reproducible, otherwise the entire simulation becomes non-reproducible. Even the first point assumption can be broken, but the simulation still has to 'survive' (with the associated errors of course)

Reply:
I've implemented a scheme that works at warp level without atomics (and without shared variables). You can launch the kernel with more than one warp (and each of them will relocate one track) or incomplete warps, but `relocateThreads  = 32` still results in the best performance.